<!-- Example: roslaunch esvo_core system_upenn.launch 
				TS_start:=0 Event_start:=10000 Dataset_Name:=upenn
				Sequence_Name:=indoor_flying1 Representation_Name:=TS -->

<launch>
	<rosparam param="/use_sim_time">true</rosparam>
	<arg name="tracking_rate_hz" default="100" />
	<arg name="Dataset_Name" default="upenn" />
	<arg name="Sequence_Name" default="indoor_flying1" />
	<arg name="Representation_Name" default="TS" />
	<arg name="eventNum_EM" default="2000" />
	<arg name="kernelSize" default="5" />

	<!-- Time surfaces generation -->
	<node name="TimeSurface_left" pkg="esvo_time_surface" type="esvo_time_surface">
		<remap from="events" to="/davis/left/events" />
		<remap from="image" to="/davis/left/image_raw" />
		<remap from="camera_info" to="/davis/left/camera_info" />
		<remap from="time_surface" to="TS_left" />
		<rosparam command="load" file="$(find esvo_core)/cfg/time_surface/ts_parameters.yaml" />
	</node> -->

	<node name="TimeSurface_right" pkg="esvo_time_surface" type="esvo_time_surface">
		<remap from="events" to="/davis/right/events" />
		<remap from="image" to="/davis/right/image_raw" />
		<remap from="camera_info" to="/davis/right/camera_info" />
		<remap from="time_surface" to="TS_right" />
		<rosparam command="load" file="$(find esvo_core)/cfg/time_surface/ts_parameters.yaml" />
	</node>

	<node name="dvs_renderer" pkg="dvs_renderer" type="dvs_renderer" output="screen" required="false">
		<remap from="events" to="/davis/left/events" />
		<remap from="image" to="/davis/left/image_raw" />
		<remap from="dvs_rendering" to="dvs_rendering" />
	</node>

	<!-- Calibration folder -->
	<arg name="calibInfoDirStr" default="$(find esvo_core)/calib/upenn"/>

	<!-- Mapping node -->
	<!-- launch-prefix="gdb -ex run -\-args" -->
	<!-- <node name="esvo_Mapping" pkg="esvo_core" type="esvo_Mapping" output="screen" required="true">
		<remap from="time_surface_left" to="/TS_left" />
		<remap from="time_surface_right" to="/TS_right" />
		<remap from="stamped_pose" to="/esvo_tracking/pose_pub" />
		<remap from="events_left" to="/davis/left/events" />
		<remap from="events_right" to="/davis/right/events" />
		<rosparam param="dvs_frame_id">"dvs"</rosparam>
		<rosparam param="world_frame_id">"map"</rosparam>
		<rosparam param="calibInfoDir" subst_value="true">$(arg calibInfoDirStr)</rosparam>
		<rosparam command="load" file="$(find esvo_core)/cfg/mapping/mapping_upenn.yaml" />
	</node> -->

	<node name="esvo_Mapping_DM" pkg="esvo_core" type="esvo_Mapping_DM" output="screen" required="true">
		<remap from="time_surface_left" to="/TS_left" />
		<remap from="time_surface_right" to="/TS_right" />
		<remap from="stamped_pose" to="/esvo_tracking/pose_pub" />
		<remap from="events_left" to="/davis/left/events" />
		<remap from="events_right" to="/davis/right/events" />
		<rosparam param="dvs_frame_id">"dvs"</rosparam>
		<rosparam param="world_frame_id">"map"</rosparam>
		<rosparam param="calibInfoDir" subst_value="true">$(arg calibInfoDirStr)</rosparam>
		<rosparam command="load" file="$(find esvo_core)/cfg/mapping/mapping_upenn.yaml" />
	</node>

	<!-- Tracking Node -->
	<node name="Tracking" pkg="esvo_core" type="Tracking" output="screen" required="true" >
		<remap from="time_surface_left" to="/TS_left" />
		<remap from="time_surface_right" to="/TS_right" />
		<remap from="stamped_pose" to="/esvo_tracking/pose_pub" />
		<remap from="gt_pose" to="/davis/left/pose" />
		<remap from="events_left" to="/davis/left/events" />
		<remap from="pointcloud" to="/esvo_mapping/pointcloud_local" />
		<rosparam param="dvs_frame_id">"dvs"</rosparam>
		<rosparam param="world_frame_id">"map"</rosparam>
		<rosparam param="calibInfoDir" subst_value="true">$(arg calibInfoDirStr)</rosparam>
		<rosparam command="load" file="$(find esvo_core)/cfg/tracking/tracking_upenn.yaml" />

		<param name="tracking_rate_hz" type="int" value="$(arg tracking_rate_hz)" />
		<param name="Dataset_Name" type="string" value="$(arg Dataset_Name)" />
		<param name="Sequence_Name" type="string" value="$(arg Sequence_Name)" />
		<param name="Representation_Name" type="string" value="$(arg Representation_Name)" />
		<param name="eventNum_EM" type="int" value="$(arg eventNum_EM)" />
	</node>

	<node name="TimeSurface_global_timer" pkg="esvo_time_surface" type="TimeSurface_global_timer" required="true" >
		<remap from="events" to="/davis/left/events" />
		<param name="frequency_timer" type="int" value="100" /> 
	</node>

	<!-- LiDARDepthMap Node -->
	<node name="LiDARDepthMap" pkg="esvo_core" type="LiDARDepthMap" required="true" output="screen">
		<remap from="stamped_pose" to="/esvo_tracking/pose_pub" />
		<remap from="gt_pose" to="/davis/left/pose" />
		<remap from="pointcloud" to="/velodyne_point_cloud" />
		<rosparam param="dvs_frame_id">"dvs"</rosparam>
		<rosparam param="lidar_frame_id">"velodyne"</rosparam>
		<rosparam param="world_frame_id">"map"</rosparam>
		<rosparam param="calibInfoDir" subst_value="true">$(arg calibInfoDirStr)</rosparam>	
		<rosparam command="load" file="$(find esvo_core)/cfg/lidardepthmap/lidardepthmap_upenn.yaml" />
		<param name="depthmap_rate_hz" type="int" value="10" />
	</node> 

	<node pkg="rosbag" type="play" name="play" required="true"
		args="/Monster/dataset/event_camera/upenn_stereo_ral/$(arg Sequence_Name)_edited.bag -r 1.0 --clock">
	</node>

    <!-- publish static tf -->
    <!-- x y z yaw pitch roll -->
    <!-- x y z qx qy qz qw -->
    <!-- <include file="$(find rviz_car_model)/launch/default.launch" /> -->
    <node name="static_tf_from_dvs_to_lidar" pkg="tf" type="static_transform_publisher"
        args="0.037 -0.086 -0.129 0.495 -0.483 0.511 0.510 dvs velodyne 100" /> 

	<!-- Visualization -->
	<!-- <node pkg="rqt_gui" type="rqt_gui" name="rqt_gui"
		args="-perspective-file $(find esvo_core)/esvo_system.perspective" />
	<node pkg="rviz" type="rviz" name="rviz"
		args="-d $(find esvo_core)/esvo_system.rviz" /> -->
</launch>
